projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,201 @@
+                              ads1015_setup_register, SIZEOF_ARRAY(ads1015_setup_register));
+  prv_check_status(code);
+  status_ok_or_return(code);
+  return STATUS_CODE_OK;
+}
+
+// reads the register and stores the value in the given array(2 bytes)
+static StatusCode prv_read(I2CPort i2c_port, ADS1015Address i2c_addr, uint8_t reg, uint8_t *read,
+                           size_t size) {
+  StatusCode code = i2c_write(i2c_port, i2c_addr | ADS1015_I2C_ADDRESS_GND, &reg, sizeof(reg));
+  prv_check_status(code);
+  status_ok_or_return(code);
+  code = i2c_read(i2c_port, i2c_addr | ADS1015_I2C_ADDRESS_GND, read, size);
+  prv_check_status(code);
+  status_ok_or_return(code);
+  return STATUS_CODE_OK;
+}
+
+// gets current channel based on the upper byte of config register
+static ADS1015Channel prv_get_channel(uint8_t config_register_msb) {
+  uint8_t channel = (config_register_msb >> 4) & 0x7;  // 4 <= channel <= 7
+  channel -= 4;  // adjusts the register value to channels' enum: 0 <= channel <= 3
+  return channel;
+}
+
+// switches to the the given channel by writing to config reg
+static StatusCode prv_set_channel(ADS1015Data *data, ADS1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  uint8_t config_register_msb_channels[] = { CONFIG_REGISTER_MSB_0, CONFIG_REGISTER_MSB_1,
+                                             CONFIG_REGISTER_MSB_2, CONFIG_REGISTER_MSB_3 };
+
+  prv_setup_register(data, ADS1015_ADDRESS_POINTER_CONFIG, config_register_msb_channels[channel],
+                     CONFIG_REGISTER_LSB);
+  data->current_channel = channel;
+  data->channel_enable[channel] = true;
+  return STATUS_CODE_OK;
+}
+
+// This function is registered as the callback for ALRT/RDY Pin.
+// reads and stores the conversion value in data, and switches to the next enabled channel.
+static void prv_interrupt_handler(const GPIOAddress *address, void *context) {
+  ADS1015Data *data = context;
+
+  uint8_t read_conv_register[2];
+  prv_read(data->i2c_port, data->i2c_addr, ADS1015_ADDRESS_POINTER_CONV, read_conv_register,
+           SIZEOF_ARRAY(read_conv_register));
+
+  uint8_t config_register[2];
+  prv_read(data->i2c_port, data->i2c_addr, ADS1015_ADDRESS_POINTER_CONFIG, config_register,
+           SIZEOF_ARRAY(config_register));
+
+  ADS1015Channel current_channel = prv_get_channel(config_register[0]);
+  data->channel_readings[current_channel][0] = read_conv_register[0];
+  data->channel_readings[current_channel][1] = read_conv_register[1];
+  // runs the users callback if not NULL
+  if (data->channel_callback[current_channel]) {
+    data->channel_callback[current_channel](data->ready_pin,
+                                            data->callback_context[current_channel]);
+  }
+  // Update so that the ADS1015 reads from the next channel
+  do {
+    current_channel = (current_channel + 1) % NUM_ADS1015_CHANNELS;
```

Comment:
This works, but you could use [find first set](https://en.wikipedia.org/wiki/Find_first_set) along with a bitset to avoid the need for the loop. I believe the ADC driver uses this concept.